### PR TITLE
Mailjet integration

### PR DESCRIPTION
**Pull request by Dcoutelle**
Please answer the following questions. 

| Q | A |
| --- | --- |
| Bug fix? | N |
| New feature? | Y |
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/86. |
| Related developer documentation PR URL | N |
| Issues addressed (#s or URLs) | N |
| BC breaks? | N |
| Deprecations? | N |

**Note that all new features should have a related user and/or developer documentation PR in their respective repositories.** 
#### Description:

The PR add mailjet as SMTP relay email router, and manage bounce / spam / blocked status callback.
#### Steps to test this PR:
1. Install this PR
2. Configuration > emails > selected mailjet in the drop down list
3. Add your mailjet login and password.
4. Login to your Mailjet account and go to My Account -> Event tracking (triggers)
5. On the event type list, select the one you want to link to your Mautic account
6. Supports the following webhooks: Message is Bounced, Message is Blocked, Message is Spam.
7. Fill in the webhook URL as `http://your-mautic.com/mailer/mailjet/callback`
8. Send a bounce email and check the mail was routed by Mailjet and declared as bounce.

### CODE REVIEW

File: app/bundles/EmailBundle/Controller/AjaxController.php [hunk 1]
```diff
@@ -44,7 +44,60 @@ protected function getAbTestFormAction(Request $request)
...
+
+
+
+
+
+
+
+
+
+
```
Comment (escopecz) [inline app/bundles/EmailBundle/Controller/AjaxController.php]: :smile: what's this for?

**Comment by escopecz** [comment]
Cool! Could you please review the code standards again?

**Comment by Dcoutelle** [comment]
@escopecz That should be OK now.

File: app/bundles/EmailBundle/Form/Type/ConfigType.php [hunk 2]
```diff
@@ -323,6 +324,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 "cram-md5"
             ], "config_emailconfig_mailer_transport":[
                 "mautic.transport.mandrill",
+				"mautic.transport.mailjet",
```
Comment (escopecz) [inline app/bundles/EmailBundle/Form/Type/ConfigType.php]: This row doesn't have the right indentation.

File: app/bundles/EmailBundle/Swiftmailer/Transport/MailjetTransport.php [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                }
+
+                if (isset($event['CustomID']) && $event['CustomID'] !== '') {
+                    $rows[$type]['hashIds'][$event['CustomID']] = $reason;
+                } else {
+                    $rows[$type] ['emails'] [$event ['email']] = $reason;
+				}
+			}
+		}
+		return $rows;
```
Comment (escopecz) [inline app/bundles/EmailBundle/Swiftmailer/Transport/MailjetTransport.php]: The 4 rows above doesn't have the right indentation.

File: app/bundles/InstallBundle/Configurator/Form/EmailStepType.php [hunk 4]
```diff
@@ -80,6 +80,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'choices'     => [
                     'mail'                      => 'mautic.email.config.mailer_transport.mail',
                     'mautic.transport.mandrill' => 'mautic.email.config.mailer_transport.mandrill',
+            	    'mautic.transport.mailjet' => 'mautic.email.config.mailer_transport.mailjet',
```
Comment (escopecz) [inline app/bundles/InstallBundle/Configurator/Form/EmailStepType.php]: And this row doesn't have the right indentation.

File: app/bundles/EmailBundle/Controller/PublicController.php [hunk 5]
```diff
@@ -297,8 +297,9 @@ public function mailerCallbackAction($transport)
         ignore_user_abort(true);
 
         // Check to see if transport matches currently used transport
-        $currentTransport = $this->factory->getMailer()->getTransport();
-
+        $currentTransportName=$this->factory->getParameter('mailer_transport');
```
Comment (alanhartless) [inline app/bundles/EmailBundle/Controller/PublicController.php]: This has been addressed with https://github.com/mautic/mautic/pull/2306

**Comment by Dcoutelle** [comment]
Removed conflicts with 2.1.1

File: app/bundles/EmailBundle/Controller/AjaxController.php [hunk 6]
```diff
@@ -228,6 +228,10 @@ protected function testEmailServerConnectionAction(Request $request)
                             $mailer->setHost($settings['amazon_region']);
                         }
                     }
+
+                    if ('mautic.transport.sparkpost' == $transport) {
```
Comment (that0n3guy) [inline app/bundles/EmailBundle/Controller/AjaxController.php]: Is this a typo?    I don't really know, but It says **sparkpost** but this is a mailjet pull request.

Comment (Dcoutelle) [inline app/bundles/EmailBundle/Controller/AjaxController.php]: You are right, these lines was added in 2.1.0 and removed in 2.1.1, I will remove them for this PR

**Comment by blmedia** [comment]
I would like to test this PR, because I use mailjet for Sending E-Mails, but I do not under how to checkout this version?!

**Comment by escopecz** [comment]
This might help: https://github.com/mautic/mautic#how-to-test-a-pull-request

**Comment by mmmichael** [comment]
Where is the option to set the API Key? I only have the possibility to set the username and no input field for the API Key!?

**Comment by escopecz** [comment]
I'm also not able to authenticate. There are fields to fill in username and password, but if I fill it in, I get `Failed to authenticate on SMTP server with username "[censored]@gmail.com" using 1 possible authenticators`. I didn't find any mention on how to authenticate in the linked documentation either. 

@Dcoutelle could you clear that up so we could test it?

File: app/bundles/EmailBundle/Controller/AjaxController.php [hunk 7]
```diff
@@ -246,10 +246,10 @@ protected function testEmailServerConnectionAction(Request $request)
                     if (empty($settings['password'])) {
                         $settings['password'] = $this->get('mautic.helper.core_parameters')->getParameter('mailer_password');
                     }
-                    $mailer->setUsername($settings['user']);
```
Comment (alanhartless) [inline app/bundles/EmailBundle/Controller/AjaxController.php]: Why were these removed?

**Comment by Dcoutelle** [comment]
I made an update on this pull request to be compatible with 2.2.1

**Comment by Dcoutelle** [comment]
I would like to change label of login/password for mailjet configuration.
On Mailjet website, login in named ApiKey and password is named ApiSecret, that why @mmmichael  was looking for these input.

I saw this was define in this file : https://github.com/mautic/mautic/blob/staging/app/bundles/EmailBundle/Form/Type/ConfigType.php#L404

But I didn't find a way to "only" change label without duplicate code... any idea ?

File: app/bundles/EmailBundle/Controller/AjaxController.php [hunk 8]
```diff
@@ -229,14 +229,14 @@ protected function testEmailServerConnectionAction(Request $request)
...
-                if (is_callable([$mailer, 'setApiKey'])) {
+                if (method_exists($mailer, 'setApiKey')) {
                     if (empty($settings['api_key'])) {
                         $settings['api_key'] = $this->get('mautic.helper.core_parameters')->getParameter('mailer_api_key');
                     }
                     $mailer->setApiKey($settings['api_key']);
                 }
 
-                if (is_callable([$mailer, 'setUsername']) && is_callable([$mailer, 'setPassword'])) {
+                if (method_exists($mailer, 'setUsername') && method_exists($mailer, 'setPassword')) {
```
Comment (alanhartless) [inline app/bundles/EmailBundle/Controller/AjaxController.php]: Method exists won't work because some mailers use overloading methods to set them. See https://github.com/mautic/mautic/pull/2750.

**Comment by alanhartless** [comment]
On the labels - it's not currently possible without a refactoring of existing code to maybe somehow update the label via Javascript or duplicate the fields which I'd rather not do. Eventually, I would like to see the whole mailer registration/configuration system refactored to something that is pluginable. I think for now, we'll just have to leverage documentation on what to do.

File: app/bundles/EmailBundle/Controller/AjaxController.php (hunk 8 above)
Comment (Dcoutelle) [inline app/bundles/EmailBundle/Controller/AjaxController.php]: I made this change because is_callable not working in dev environement and lead to error 500 with any mailer when you click on test button

Comment (alanhartless) [inline app/bundles/EmailBundle/Controller/AjaxController.php]: Right - that's what the PR I referenced is fixing. method_exists doesn't work for SMTP because the method doesn't exist since Swiftmailer's transport uses __call for those.

**Comment by qneyrat** [comment]
When send email with mailjet transport {contactfield} of {formfield}  is not set because after my search `bundles/EmailBundle/Helper/MailHelper.php` line 346 
`if (!$isQueueFlush)` and `if( method_exists($this->message, 'addMetadata' )` but method to fill contactfield or formfield `self::searchReplaceTokens($search, $replace, $this->message);` in else

**Comment by alanhartless** [comment]
@Dcoutelle I think the issue may be extending the `AbstractTokenSMTPTransport`. That class is meant for transports that support token replacement/substitution tags via headers so if those headers aren't defined in `prepareMessage`, they aren't replaced (contact fields, etc). 

If you want Mautic to handle the content replacement for a SMTP based mailer, then extend `extend \Swift_SmtpTransport` instead then just inherit `public function send(Swift_Mime_Message $message, &$failedRecipients = null)`, set the custom header there then execute `parent::send($message, $failedRecipients)` in place of using `prepareMessage`.

**Comment by Dcoutelle** [comment]
I updated the PR, token replacement work with immediate send also

Thanks @alanhartless for tips!

**Comment by qneyrat** [comment]
token replacement is fix, good job ! 👍

**Comment by escopecz** [comment]
@Dcoutelle Could you please merge this branch with latest staging and fix the conflicts?

**Comment by Dcoutelle** [comment]
@escopecz done!

**Comment by escopecz** [comment]
Looks good to me :+1:

**Comment by mmmichael** [comment]
Hello, I merged this request into my version 2.2.1 and everything works fine!
